File: Sources/custom_adapter/UnityLevelPlayAdapter/Objc/ISAlxCustomBanner.m
//
//  ISAlxCustomBanner.m
//  AlxAdsOCDemo
//
//  LevelPlay Banner 广告适配器
//

#import "ISAlxCustomBanner.h"
#import <AlxAds/AlxAds-Swift.h>

static NSString *const TAG = @"ISAlxCustomBanner";

@interface ISAlxCustomBanner () <AlxBannerViewAdDelegate>
@property (nonatomic, strong, nullable) AlxBannerAdView *bannerAdView;
@property (nonatomic, weak, nullable) id<ISBannerAdDelegate> adDelegate;
@end

@implementation ISAlxCustomBanner

#pragma mark - ISBaseBanner

/// LevelPlay 请求加载 Banner 广告
/// adData.configuration 包含:
///   appid / sid / token (app 级) + unitid (instance 级)
- (void)loadAdWithAdData:(ISAdData *)adData
         viewController:(UIViewController *)viewController
                   size:(ISBannerSize *)size
               delegate:(id<ISBannerAdDelegate>)delegate {
    NSLog(@"%@: loadAd", TAG);
    self.adDelegate = delegate;

    [ISAlxCustomAdapter initSdkWithAdData:adData];

    NSString *unitId = adData.configuration[@"unitid"];
    if (!unitId || unitId.length == 0) {
        NSString *msg = @"unitid is empty";
        NSLog(@"%@: error: %@", TAG, msg);
        [delegate adDidFailToLoadWithErrorType:ISAdapterErrorTypeInternal
                                     errorCode:ISAdapterErrorMissingParams
                                  errorMessage:msg];
        return;
    }

    NSLog(@"%@: loadAd unitid=%@", TAG, unitId);

    CGSize adSize = [self bannerSizeFromISSize:size];
    dispatch_async(dispatch_get_main_queue(), ^{
        self.bannerAdView = [[AlxBannerAdView alloc] initWithFrame:CGRectMake(0, 0, adSize.width, adSize.height)];
        self.bannerAdView.delegate = self;
        self.bannerAdView.refreshInterval = 0;
        self.bannerAdView.rootViewController = viewController;
        [self.bannerAdView loadAdWithAdUnitId:unitId];
    });
}

/// LevelPlay 销毁 Banner
- (void)destroyAdWithAdData:(ISAdData *)adData {
    NSLog(@"%@: destroyAd", TAG);
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.bannerAdView destroy];
        self.bannerAdView = nil;
    });
}

#pragma mark - AlxBannerViewAdDelegate

- (void)bannerViewAdLoad:(AlxBannerAdView *)bannerView {
    NSLog(@"%@: bannerViewAdLoad", TAG);
    [self.adDelegate adDidLoadWithView:bannerView];
}

- (void)bannerViewAdFailToLoad:(AlxBannerAdView *)bannerView didFailWithError:(NSError *)error {
    NSLog(@"%@: bannerViewAdFailToLoad: %@", TAG, error.localizedDescription);
    [self.adDelegate adDidFailToLoadWithErrorType:ISAdapterErrorTypeInternal
                                        errorCode:error.code
                                     errorMessage:error.localizedDescription];
}

- (void)bannerViewAdImpression:(AlxBannerAdView *)bannerView {
    NSLog(@"%@: bannerViewAdImpression", TAG);
    [self.adDelegate adDidOpen];
}

- (void)bannerViewAdClick:(AlxBannerAdView *)bannerView {
    NSLog(@"%@: bannerViewAdClick", TAG);
    [self.adDelegate adDidClick];
}

- (void)bannerViewAdClose:(AlxBannerAdView *)bannerView {
    NSLog(@"%@: bannerViewAdClose", TAG);
    [self.adDelegate adDidDismissScreen];
}

#pragma mark - Private

/// 将 LevelPlay ISBannerSize 映射为 CGSize
/// ⚠️ 此方法可能在后台线程调用，禁止访问 vc.view（UI API），改用 UIScreen
- (CGSize)bannerSizeFromISSize:(ISBannerSize *)size {
    CGFloat screenWidth = UIScreen.mainScreen.bounds.size.width;

    if ([size.sizeDescription isEqualToString:@"RECTANGLE"]) {
        return CGSizeMake(300, 250);
    } else if ([size.sizeDescription isEqualToString:@"LARGE"]) {
        return CGSizeMake(320, 90);
    } else if ([size.sizeDescription isEqualToString:@"SMART"] || [size.sizeDescription isEqualToString:@"BANNER"]) {
        return CGSizeMake(screenWidth, 50);
    }
    return CGSizeMake(320, 50);
}

@end

File: custom_adapter/UnityLevelPlayAdapter/Swift/ISAlxCustomBanner.swift
//
//  ISAlxCustomBanner.swift
//  AlxAdsDemo
//
//  LevelPlay Banner 广告适配器 / LevelPlay Banner Ad Adapter
//  文档参考 / Documentation: https://docs.unity.com/zh-cn/grow/levelplay/sdk/ios/build-custom-adapter
//

import Foundation
import UIKit
import IronSource
import AlxAds

@objc(ISAlxCustomBanner)
public class ISAlxCustomBanner: ISBaseBanner {

    private static let TAG = "ISAlxCustomBanner"

    private var bannerAdView: AlxBannerAdView?
    private weak var adDelegate: ISBannerAdDelegate?

    // MARK: - ISBaseBanner

    /**
     * LevelPlay 请求加载 Banner 广告。
     * LevelPlay requests to load a Banner ad.
     *
     * adData.configuration 包含 / contains:
     *   appid / sid / token (app 级 / app-level) + unitid (instance 级 / instance-level)
     */
    public override func loadAd(with adData: ISAdData,
                                viewController: UIViewController,
                                size: ISBannerSize,
                                delegate: ISBannerAdDelegate) {
        NSLog("%@: loadAd", Self.TAG)
        self.adDelegate = delegate

        ISAlxCustomAdapter.initSdk(with: adData)

        guard let unitId = adData.configuration["unitid"] as? String, !unitId.isEmpty else {
            let msg = "unitid is empty"
            NSLog("%@: error: %@", Self.TAG, msg)
            (delegate as? ISAlxAdapterFailureReporter)?.adDidFailToLoad(
                withErrorType: 2,
                errorCode: ISAdapterErrors.missingParams.rawValue,
                errorMessage: msg)
            return
        }

        NSLog("%@: loadAd unitid=%@", Self.TAG, unitId)

        let adSize = bannerSize(from: size)
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            let banner = AlxBannerAdView(frame: CGRect(origin: .zero, size: adSize))
            banner.delegate = self
            banner.refreshInterval = 0
            banner.rootViewController = viewController
            self.bannerAdView = banner
            banner.loadAd(adUnitId: unitId)
        }
    }

    /**
     * LevelPlay 销毁 Banner。
     * LevelPlay destroys the Banner.
     */
    public override func destroyAd(with adData: ISAdData) {
        NSLog("%@: destroyAd", Self.TAG)
        DispatchQueue.main.async { [weak self] in
            self?.bannerAdView?.destroy()
            self?.bannerAdView = nil
        }
    }

    // MARK: - Private

    /**
     * 将 LevelPlay ISBannerSize 映射为 CGSize。
     * Maps LevelPlay ISBannerSize to CGSize.
     *
     * ⚠️ 此方法可能在后台线程调用，禁止访问 vc.view（UI API），改用 UIScreen。
     * ⚠️ This method may be called on a background thread; do not access vc.view (UI API), use UIScreen instead.
     */
    private func bannerSize(from size: ISBannerSize) -> CGSize {
        let screenWidth = UIScreen.main.bounds.size.width
        switch size.sizeDescription {
        case "RECTANGLE":
            return CGSize(width: 300, height: 250)
        case "LARGE":
            return CGSize(width: 320, height: 90)
        case "SMART", "BANNER":
            return CGSize(width: screenWidth, height: 50)
        default:
            return CGSize(width: 320, height: 50)
        }
    }
}

// MARK: - AlxBannerViewAdDelegate

extension ISAlxCustomBanner: AlxBannerViewAdDelegate {

    public func bannerViewAdLoad(_ bannerView: AlxBannerAdView) {
        NSLog("%@: bannerViewAdLoad", ISAlxCustomBanner.TAG)
        adDelegate?.adDidLoad(with: bannerView)
    }

    public func bannerViewAdFailToLoad(_ bannerView: AlxBannerAdView, didFailWithError error: Error) {
        NSLog("%@: bannerViewAdFailToLoad: %@", ISAlxCustomBanner.TAG, error.localizedDescription)
        let nsError = error as NSError
        (adDelegate as? ISAlxAdapterFailureReporter)?.adDidFailToLoad(
            withErrorType: 2,
            errorCode: nsError.code,
            errorMessage: error.localizedDescription)
    }

    public func bannerViewAdImpression(_ bannerView: AlxBannerAdView) {
        NSLog("%@: bannerViewAdImpression", ISAlxCustomBanner.TAG)
        adDelegate?.adDidOpen()
    }

    public func bannerViewAdClick(_ bannerView: AlxBannerAdView) {
        NSLog("%@: bannerViewAdClick", ISAlxCustomBanner.TAG)
        adDelegate?.adDidClick()
    }

    public func bannerViewAdClose(_ bannerView: AlxBannerAdView) {
        NSLog("%@: bannerViewAdClose", ISAlxCustomBanner.TAG)
        adDelegate?.adDidDismissScreen()
    }
}
